File: 프로그래머스/lv2/12913. 땅따먹기/땅따먹기.js
function solution(land) {
    return maxSum(land);
}

const maxSum = land => {
    const [row, col] = [land.length, land[0].length];
    
    const dp = Array.from(Array(row), () => Array(col).fill(0));
        dp[0][0] = land[0][0];
        dp[0][1] = land[0][1];
        dp[0][2] = land[0][2];
        dp[0][3] = land[0][3];
    
    // // 첫줄 채우기
    // for(let i = 0; i < col; i++) {
    //     dp[0][i] = land[0][i];
    // }
    
    for(let i = 1; i < row; i++) {
        
        for(let j = 0; j < col; j++) {
            const rowArr = [];
            
            for(let k = 0; k < col; k++) {
                if(k !== j) rowArr.push(dp[i-1][k]);
            }
            
            dp[i][j] = Math.max(...rowArr) + land[i][j];
        }
    }
    
    return Math.max(...dp[row-1]);
}

/*
2차원 배열을 선언해야한다. 배열의 행과 열은 land와 똑같이 한다. 
for를 두번 반복해서 각 칸에서의 최대 max를 구한다. 
마지막 행에서 최대값을 반환한다. 
*/

File: 프로그래머스/2/12913. 땅따먹기/땅따먹기.py
'''
누적하는 문제. 
위 아래 크기가 모두 같으므로 어디서 시작하든 똑같음
'''

def solution(land):
    s = [[0 for _ in range(len(land[0]))] for _ in range(len(land))]
    for i in range(len(land[0])):
        s[0][i] = land[0][i]
        
    for y in range(1, len(land)):
        for x in range(len(land[y])):
            for i in range(len(land[y])):
                if i != x:
                    s[y][x] = max(s[y][x], s[y - 1][i] + land[y][x])
                    
    return max(s[-1])
            
